Preserve Union type annotations during serialization

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58911
* #58673
* #57014
* __->__ #57013
* #57012
* #57011
* #53180

Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

Commit:
Preserve Union type annotations during serialization

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"




[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

eellison reviewed (commented):
@gmagogsfm mind reviewing this one ?

gmagogsfm commented:
> @gmagogsfm mind reviewing this one ?

I will get it reviewed today

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -510,13 +510,36 @@ struct PythonPrintImpl {
+    if (lhs->type()->kind() == UnionType::Kind ||
+        lhs->type()->kind() == OptionalType::Kind) {
+      if (!annotated_unions_.count(lhs)) {
+        annotated_unions_.insert(lhs);
```
Optional Nit:

`insert` will actually return a pair of values, second of which will tell you whether an insertion actually took place or not. If you use that value, you can save one lookup in `count` and logic will be simpler.

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -510,13 +510,36 @@ struct PythonPrintImpl {
   bool requiresAnnotation(Value* lhs, Value* rhs) {
-    return *lhs->type() != *rhs->type();
+    if (lhs->type()->kind() == UnionType::Kind ||
+        lhs->type()->kind() == OptionalType::Kind) {
```
Good catch! I wonder how this never caught on fire for `OptionalType`

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -510,13 +510,36 @@ struct PythonPrintImpl {
     indent();
     printValueList(body_, lhs);
+    // We need to preserve Union/Optional type annotations, but only if
+    // we're not assigning values as part of a tuple unpacking statement
```
Does this mean unpacking tuple into a `union`-typed variable is not supported now? If so, let's add an explicit error and a `TODO` for future fix.

ansley replied:
Can you give an example? We only add annotations on the first assignment of a variable, and I can't think of a case in which some lhs could be a `Union` without an annotation. (Obviously, having two Union annotations on the same line is a SyntaxError!) Thank you!

gmagogsfm replied:
Thanks for clarification, I didn't know about this Python limitation before.

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28572813).

Commit:
Update on "Preserve Union type annotations during serialization"


Differential Revision: [D28572813](https://our.internmc.facebook.com/intern/diff/D28572813)

[ghstack-poisoned]
